angular/lib/src/core/change_detection/differs/default_iterable_differ.dart

```diff
@@ -13,9 +13,32 @@ typedef void DefaultIterableCallback(
   int currentIndex,
 );
 
-typedef dynamic TrackByFn(int index, dynamic item);
-
-var trackByIdentity = (int index, dynamic item) => item;
+/// A function that can be used to return a unique key for [item] at [index].
+///
+/// By default, [item] itself is used as the key to track to instantiate a new
+/// template per item in `*ngFor`. If the data ever changes, AngularDart will
+/// consider any object with a different identity (`identical`) to be different
+/// - and destroy and re-create a new node.
+///
+/// To optimize performance when you have a way to determine uniqueness other
+/// than identity, such as an `id` field on an object returned from the server,
+/// you may specify a [TrackByFn]`:
+/// ```
+/// class MyComp {
+///   Object trackByEmployeeId(int index, Object item) => (item as Employee).id;
+/// }
+///
+/// class Employee {
+///   int id;
+/// }
+/// ```
+///
+/// **NOTE**: It is not safe to simply _assume_ that the second parameter is
+/// of your custom type (neither [TrackByFn] nor `NgFor` allow that) at this
+/// time: https://github.com/dart-lang/angular/issues/1020.
+typedef Object TrackByFn(int index, Object item);
```

Comment:
Just curious: why change the type of `item` from `dynamic` to `Object`? Essentially, the trackBy function is the user suggested way of comparing values for equality, and the `Object.==` is typed as: `bool operator ==(other);`

In terms of #1020, are you thinking of something like this?

```dart
typedef TrackByFn<T,E> = T Function(int index, E item);
```

Reply:
> Just curious: why change the type of `item` from `dynamic` to `Object`? Essentially